---
title: "Amplicon_analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set work directory
```{r, message=FALSE, warning=TRUE}
setwd('e:/permafrost/')
wd_16s <- file.path(getwd(),"data/16S")
# if (!dir.exists(wd_16s)) {
#   dir.create(wd_16s)
# }
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r, message=FALSE, warning=FALSE}
library(phyloseq)
library(ape)
library(vegan)
library(Biostrings)
library(microbiome)
library(tidyverse)
```
Data input
```{r, message=FALSE, warning=TRUE}
source("script/read_data.R")
```

## Alpha diversity
Determine the alpha diversity including **Observed**, **Chao1**, **Shannon** and **Simpson**.
```{r, message=FALSE, warning=TRUE}
alpha_div <- estimate_richness(phylo_rare, measures = c("Observed", "Chao1", 'Shannon', 'Simpson'))
library(picante)
pd <- pd(t(otu), tree, include.root = F) # estimate the phylogenetic diversity
alpha_div <- cbind(Layers = metadata$Layer, alpha_div, Faith = pd$PD) %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL')))
```
Box plot for evenness index using ggplot2
```{r, message=FALSE, warning=TRUE}
library(ggpubr)
library(ggplot2)
my_comparisons <- list( c("SUR", "SUB"), c("SUB", "PL"), c("SUR", "PL"))
p_shannon <- ggplot(alpha_div, aes(x = Layers, y = Shannon)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons, paired = TRUE, 
                     p.adjust.method = "BH", label = "p.signif") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'Shannon', fill='Layers') +
  theme_bw() +
  theme(axis.title = element_text(size = 12, colour = "black"),
        axis.text = element_text(size = 10, colour = "black"),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10),
        panel.grid = element_blank())
```
Box plot for faith index using ggplot2
```{r, message=FALSE, warning=TRUE}
p_faith <- ggplot(alpha_div, aes(x = Layers, y = Faith)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons,  paired = TRUE, 
                     p.adjust.method = "BH", label = "p.signif") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'Faith index', fill='Layers') +
  theme_bw() +
  theme(axis.title = element_text(size = 12, colour = "black"),
        axis.text = element_text(size = 10, colour = "black"),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10),
        panel.grid = element_blank())
```
Then, we combine above two alpha diversity plot into one.
```{r, fig.align='center', fig.width=6.5, fig.height=3.5}
cowplot::plot_grid(p_shannon, p_faith)
```
## Beta diversity
```{r, message=FALSE, warning=FALSE}
library(vegan)
## taxonomy
### PERMANOVA test
env_dist <- as.matrix(vegdist(scale(metadata[,-c(1:4)]), "euclidean" ))
tax_dist <- as.matrix(vegdist(t(otu), "bray" ))
phy_dist <- as.matrix(UniFrac(phylo, weighted = TRUE, normalized = TRUE, parallel = T, fast = TRUE))
fun_dist <- as.matrix(vegdist(t(ko_tpm_table), "bray" ))
```

```{r, message=FALSE, warning=TRUE}
#permanova, ANOSIM and MRPP analysis
library(vegan)
adonis2(tax_dist ~ Layer, data = metadata)
adonis2(phy_dist ~ Layer, data = metadata)
adonis2(fun_dist ~ Layer, data = metadata)

mrpp(tax_dist, metadata$Layer, perm = 999)
mrpp(phy_dist, metadata$Layer, perm = 999)
mrpp(fun_dist, metadata$Layer, perm = 999)

anosim(tax_dist, metadata$Layer, perm = 999)
anosim(phy_dist, metadata$Layer, perm = 999)
anosim(fun_dist, metadata$Layer, perm = 999)
```

### PCoA plot with bray-curties as distance
```{r, fig.align='center', fig.width=8, fig.height=6}
# taxa
ord.tax <-  cmdscale(tax_dist,  k = 2, eig = T, add = T)
pcoa_tax_plot <- data.frame(Layers = metadata$Layer, scores(ord.tax)) %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Layers, color = Layers)) + 
  geom_point(size = 1, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Layers), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  labs(x=paste("PCoA1 (", format(100 * ord.tax$eig[1] / sum(ord.tax$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.tax$eig[2] / sum(ord.tax$eig), digits = 3), "%)", sep = "")) +
  theme(axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
# ggsave(file.path(save.dir, "./figs/beta/PCoA_tax_bray.pdf"),
#        pcoa_tax_plot, width = 89, height = 59, units = "mm")

# phylogency
ord.phy <- ordinate(phylo, method = "PCoA", distance = "unifrac", weighted = TRUE)
pcoa_unifrac_plot <- data.frame(Layers = metadata$Layer, ord.phy$vectors[, 1:2]) %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(x = Axis.1, y = Axis.2, shape = Layers, color = Layers)) + 
  geom_point(size = 1, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Layers), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  labs(x=paste("PCoA1 (", format(100 * ord.phy$values[1, 2], digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.phy$values[2, 2], digits = 3), "%)", sep = "")) +
  theme(axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
# ggsave(file.path(save.dir, "./figs/beta/PCoA_unifrac_plot.pdf"),
#        PCoA_unifrac_plot, width = 89, height = 59, units = "mm")

# function
ord.fun <-  cmdscale(fun_dist,  k = 2, eig = T, add = T)
library(ggplot2)
pcoa_fun_plot <- data.frame(Layers = metadata$Layer, scores(ord.fun)) %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Layers, color = Layers)) + 
  geom_point(size = 1.5, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Layers), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  labs(x=paste("PCoA1 (", format(100 * ord.fun$eig[1] / sum(ord.fun$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.fun$eig[2] / sum(ord.fun$eig), digits = 3), "%)", sep = "")) +
  theme(axis.title = element_text(size = 6, colour = "black"),
        axis.text = element_text(size = 5, colour = "black"),
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 5),
        legend.key = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
# ggsave(file.path(save.dir, "./figs/beta/PCoA_fun_bray.pdf"),
#        pcoa_fun_plot, width = 89, height = 59, units = "mm")

# arrange the PCoA plot
cowplot::plot_grid(pcoa_tax_plot, pcoa_unifrac_plot, pcoa_fun_plot)
```

### Difference in taxonomic variance among layers
```{r}
beta_tax_plot <- sapply(unique(metadata$Layer), function(x) usedist::dist_subset(tax_dist, grep(x, metadata$sample_id, value = T))) %>%
  data.frame() %>% gather("Layers", "distance") %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(x = Layers, y = distance)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons,  paired = TRUE, 
                     p.adjust.method = "BH") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'Taxonomic variance', fill='Layers') +
  theme_bw() +
  theme(axis.title = element_text(size = 12, colour = "black"),
        axis.text = element_text(size = 10, colour = "black"),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10),
        panel.grid = element_blank())
# ggsave(file.path(save.dir, "./figs/beta/tax_variance.pdf"),
#        beta_tax_plot, width = 89, height = 89, units = "mm")
```

### Difference in phylogenetic variance among layers
First, we determine the betaMNTD using picante package.
```{r}
require(picante)
#Time consuming
# beta.mntd.weighted <- as.matrix(comdistnt(t(data.frame(otu)), cophenetic(tree), abundance.weighted = T))

beta.mntd.weighted <- read.table(file.path(save.dir, './tables/beta.mntd.weighted.txt'),
                                 header = T, row.names = 1)
beta_MNTD_plot <- sapply(unique(metadata$Layer), function(x) usedist::dist_subset(beta.mntd.weighted, grep(x, metadata$sample_id, value = T))) %>%
  data.frame() %>% gather("Layers", "distance") %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(x = Layers, y = distance)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons,  paired = TRUE, 
                     p.adjust.method = "BH") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'beta-MNTD', fill='Layers') +
  theme_bw() +
  theme(axis.title = element_text(size = 12, colour = "black"),
        axis.text = element_text(size = 10, colour = "black"),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10),
        panel.grid = element_blank())
# ggsave(file.path(save.dir, "./figs/beta/beta_mntd.pdf"),
#        beta_MNTD_plot, width = 89, height = 89, units = "mm")
```

### Difference in functional variance among layers
```{r}
beta_fun_plot <- sapply(unique(metadata$Layer), function(x) usedist::dist_subset(fun_dist, grep(x, metadata$sample_id, value = T))) %>%
  data.frame() %>% gather("Layers", "distance") %>%
  mutate(Layers = factor(Layers, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(x = Layers, y = distance)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons,  paired = TRUE, 
                     p.adjust.method = "BH") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'Functional variance', fill='Layers') +
  theme_bw() +
  theme(axis.title = element_text(size = 12, colour = "black"),
        axis.text = element_text(size = 10, colour = "black"),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10),
        panel.grid = element_blank())
# ggsave(file.path(save.dir, "./figs/beta/fun_variance.pdf"),
#        beta_fun_plot, width = 89, height = 89, units = "mm")
beta_fun_plot
```

Then, we arrange the plot using cowplot package
```{r, fig.align='center', fig.width=8, fig.height=3.5}
cowplot::plot_grid(beta_tax_plot, beta_MNTD_plot)
```

## Composition
### Determine the OTU numbers within each dominant phylum
```{r}
# data preparation
otu_num_in_phylum <- data.frame(table(tax_table(phylo)[,"Phylum"]))
otu_num_in_phylum <- otu_num_in_phylum %>% arrange(desc(Freq))
otu_num_in_phylum <- rbind(otu_num_in_phylum[1:11, ], data.frame(Var1 = c('Others'), Freq = sum(otu_num_in_phylum[-c(1:11), 2])))

otu_num_in_phylum <- data.frame(Phylum = otu_num_in_phylum$Var1, otu_num = otu_num_in_phylum$Freq ,
                                prop = otu_num_in_phylum$Freq/sum(otu_num_in_phylum$Freq)*100)
otu_count.data <- otu_num_in_phylum %>%
  mutate(lab.ypos = cumsum(prop) - 0.5*prop)
otu_count.data$Phylum <- factor(otu_count.data$Phylum, ordered = T, levels = otu_num_in_phylum$Phylum)
```

```{r}
# pie plot 
# Define the colors you want
mycols <- c("#89c5da", "#ffc15c", "#74d944", "#CE50CA", "#5e738f", "#C0717C", "#CBD5ec", "#5F7FC7", 
                     "#00718b", "#00b0f6", "#a3a500", "#f8766d", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD", 
                     "#D14285", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F", "#D1A33D", 
                     "#8A7C64", "#599861")

pie_for_otu_num_phylum <- ggplot(otu_count.data, aes(x = "", y = prop, 
                                                     fill = reorder(Phylum, -lab.ypos))) +
  geom_bar(width = 1, stat = "identity") + 
  coord_polar("y", start = 0) +
  geom_text(aes(x = 1.35, y = lab.ypos, 
                label = paste0(otu_num, ' (', round(prop, 1), '%', ')', sep = '')),
            color = "black", size = 3) +
  scale_fill_manual('Phylum', values = mycols) +
  guides(fill = guide_legend(reverse = T)) +
  theme_void() +
  theme(legend.position = "left")
```

### Determine the microbial composition within top 10 phyla
```{r}
# write a function for data organization
arrange.tab <- function(phylo, N, taxrank, vect) {
  subphylo <- tax_glom(phylo, taxrank)
  subphylo.rel <- microbiome::transform(subphylo, "compositional")
  ra.tab <- otu_table(subphylo.rel)
  MRA <- rowMeans(ra.tab)
  group <- tax_table(subphylo.rel)[,vect]
  mra.tab <- data.frame(group,MRA)
  colnames(mra.tab) <- c('level1', 'level2', 'MRA')
  #arrange the class table
  mra.tab_level1 = mra.tab %>% group_by(level1) %>% 
    summarise(sum_MRA = sum(MRA)) %>% 
    arrange(desc(sum_MRA))
  top_N_level1 = mra.tab_level1[1:N, ]$'level1'
  top_N_tab = mra.tab[mra.tab$'level1' %in% top_N_level1, ]
  mra.tab_level2 = top_N_tab %>% group_by(level2) %>% 
    summarise(sum_MRA = sum(MRA)) %>% 
    arrange(desc(sum_MRA))
  order_level2 = mra.tab_level2$'level2'
  top_N_tab$'level1' = factor(top_N_tab$'level1', ordered = T, levels = top_N_level1)
  top_N_tab$'level2' = factor(top_N_tab$'level2', ordered = T, levels = rev(order_level2))
  top_N_tab$MRA = top_N_tab$MRA*100
  return(top_N_tab)
}
```
Determine the relative abundance of the taxa at the phylum level.
```{r}
top10phylum <- arrange.tab(phylo, 10, 'Order', c(2,4))
mra.tab_level1 = top10phylum %>% group_by(level1) %>% 
  summarise(sum_MRA = sum(MRA)) %>% 
  arrange(desc(sum_MRA)) %>%
  dplyr::rename(Phylum = level1, prop = sum_MRA) %>%
  bind_rows(summarise_all(., ~if(is.numeric(.)) 100-sum(.) else "Others"))
mra.tab_level1
```
Pie plot for the relative abundance of dominant phyla
```{r, fig.align='center', fig.width=8, fig.height=4.5}
rel_abun_phyla_tab <- mra.tab_level1 %>%
  mutate(lab.ypos = cumsum(prop) - 0.5*prop)
rel_abun_phyla_tab$Phylum <- factor(rel_abun_phyla_tab$Phylum, 
                                ordered = T, levels = c(mra.tab_level1$Phylum))

pie_for_abun_phylum <- ggplot(rel_abun_phyla_tab, aes(x = "", y = prop, 
                                                     fill = reorder(Phylum, -lab.ypos))) +
  geom_bar(width = 1, stat = "identity") + 
  coord_polar("y", start = 0) +
  geom_text(aes(x = 1.35, y = lab.ypos, label = paste0(round(prop, 1), '%', sep= '')),
            color = "black", size = 3) +
  scale_fill_manual('Phylum', values = mycols[-1]) +
  guides(fill = guide_legend(reverse = T)) +
  theme_void() +
  theme(legend.position = "left")
cowplot::plot_grid(pie_for_otu_num_phylum, pie_for_abun_phylum)
```

Determine the relative abundance of the taxa at the order level.
```{r}
mra.tab_level2 = top10phylum %>% group_by(level2) %>% 
  summarise(sum_MRA = sum(MRA)) %>% 
  arrange(desc(sum_MRA))
order_level2 = mra.tab_level2$'level2'
mra.tab_level2 [1:20, ]
```
### Determine the compositions within top 10 phylums for each layer
```{r}
subphylo <- tax_glom(phylo, 'Phylum')
subphylo.rel  = transform_sample_counts(subphylo, function(x) x / sum(x) )
ntaxa(subphylo.rel)
ra.tab <- otu_table(subphylo.rel)
sum(ra.tab[, 1])
subtaxa_tab <- tax_table(subphylo.rel)[, 2]
Phylum_tax_table <- data.frame(subtaxa_tab, ra.tab) %>% 
  mutate(MRA = rowMeans(.[, colnames(ra.tab)])) %>%
  arrange(desc(MRA)) %>% dplyr::top_n(10, MRA) %>% 
  dplyr::select(-MRA) %>% 
  bind_rows(summarise_all(., ~if(is.numeric(.)) 1-sum(.) else "Others")) %>%
  mutate(Phylum = factor(Phylum, levels = Phylum)) %>%
  pivot_longer(cols = -c(Phylum), names_to = "Sample_name", values_to = 'rel_abun') %>%
  right_join(data.frame(Sample_name = metadata$sample_id, Layer = metadata$Layer), by = c("Sample_name")) %>%
  dplyr::select(-Sample_name) %>% 
  group_by(Layer, Phylum) %>%
  dplyr::summarise(across(, mean, na.rm = TRUE))
print(Phylum_tax_table, n = 32)
```

```{r, fig.width = 8, fig.height = 4.5}
## boxplot shows the community composition for each layer
box_plot <- data.frame(subtaxa_tab, ra.tab) %>% group_by(Phylum) %>% 
  summarise(across(everything(), sum)) %>% 
  mutate(MRA = rowMeans(.[, colnames(ra.tab)])) %>%
  arrange(desc(MRA)) %>% dplyr::top_n(10, MRA) %>%
  dplyr::select(-MRA) %>% 
  mutate(Phylum = factor(Phylum, levels = Phylum)) %>%
  tidyr::pivot_longer(cols = -c(Phylum), names_to = "sample_id", values_to = 'rel_abun') %>%
  mutate(layer = sapply(stringr::str_split(sample_id, "_",  n = 2), `[`, 1)) %>%
  mutate(layer = factor(layer, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(Phylum, rel_abun*100)) + 
  geom_boxplot(width = 0.5, aes(fill = layer)) +
  stat_compare_means(aes(group = layer),  paired = TRUE, 
                     p.adjust.method = "BH", label = "p.signif") +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  labs(x = 'Phylum', y = 'Relative abundance (%)', fill='Layers') +
  theme_bw() +
  guides(fill = guide_legend(keywidth = 0.5, keyheight = 0.5)) +
  theme(axis.title = element_text(size = 8, colour = "black"),
        axis.text.x = element_text(size = 6, colour = "black", 
                                   angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank())
box_plot
```

```{r, fig.width = 8, fig.height = 6}
# Barplot shows the community composition for each layer
# bar plot at the Phylum level
bar_plot_each_layer <- Phylum_tax_table %>%
  ggplot(aes(x = Layer, y = 100*rel_abun, fill = Phylum))+
  geom_bar(stat = "identity") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(limits = rev(c('SUR', 'SUB', 'PL'))) +
  scale_fill_manual(values =  mycols) +
  labs(x = 'Layer', y = 'Mean relative abundance (%)') +
  guides(fill = guide_legend(keywidth = 0.5, keyheight = 0.5)) +
  theme_linedraw() + 
  theme(panel.grid=element_blank(), 
        strip.text = element_text(colour = 'black', size = 8),
        strip.background = element_rect(colour = 'grey', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 8),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color = 'black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour = 'black', size = 8),
        axis.text.x = element_text(colour = 'black', size = 8),
        legend.position = "bottom",
        legend.title = element_blank(),
        legend.text = element_text(size = 8),
        legend.key = element_blank(),
        legend.background = element_rect(colour = "white")) +
  guides(fill = guide_legend(ncol = 5)) +
  coord_flip()

library(cowplot)
compositional_plot <- ggdraw() +
  draw_plot(pie_for_otu_num_phylum, x = 0, y = 1/2, width = 0.5, height = 1/2) +
  draw_plot(pie_for_abun_phylum, x = 0.5, y = 1/2, width = 0.5, height = 1/2) +
  draw_plot(bar_plot_each_layer, x = 0, y = 0, width = 1, height = 1/2) +
  draw_plot_label(label = c("A", "B", "C"), size = 8,
                  x = c(0, 0.5, 0), y = c(1, 1, 0.5))
compositional_plot
```

### Unique otus profile among three layers
```{r, fig.width = 6, fig.height = 4.5}
# library
library(ggvenn)
#Make the plot
x = list(
  SUR = otu %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('SUR', metadata$sample_id, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames(),
  SUB = otu %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('SUB', metadata$sample_id, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames(),
  PL = otu %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('PL', metadata$sample_id, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames()
  )

ggvenn(
  x, 
  fill_color = c("#f8766d", "#a3a500", "#00b0f6"),
  stroke_color = NA,
  set_name_size = 4,
  text_size = 4,
  show_percentage = F
)
```

