---
title: "Analysis for revision"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set work and saving directory
```{r}
wd <- 'e:/permafrost'
setwd(wd)
save.dir.revision <- file.path(wd,"revision/results")
if (!dir.exists(save.dir.revision)) {
  dir.create(save.dir.revision)
}
```

## Reading data
```{r}
HF_df <- read.table(file = file.path(wd, 'revision/data/HF.csv'),
                           sep = ',',  header = T, stringsAsFactors = F)
```
Test the human disturbance on the Tibetan plateau
```{r}
# plot
library(gghalves)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
my_comparisons <- list(c('Sampling_sites', 'Cities'))
p1 <- HF_df %>%
  mutate(Type = factor(Type, levels = c("Sampling_sites", "Cities"))) %>%
  ggplot(aes(Type, HF_2000_2015, fill = Type)) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Type), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_x_discrete(labels = c("Sampling_sites" = "Sampling sites", "Cities" = "Cities")) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = "Human footprint index") +
  scale_fill_manual(values = c("#f8766d", "#a3a500")) +
  main_theme + theme(legend.position = "none")
# save plot
# ggsave(file.path(save.dir.revision, "./HF.pdf"), p1, width = 3.5, height = 2.5, units = "in")
p1
```

## Explor the phylogenecy of specific taxonomic groups and comparing the phylogenetic difference among layers
```{r}
# First, we want to know which taxonomic groups are dominant taxa in our samples


```

```{r}
# write a function for data organization
arrange.tab <- function(phylo, N, taxrank, vect) {
  subphylo <- tax_glom(phylo, taxrank)
  subphylo.rel <- microbiome::transform(subphylo, "compositional")
  ra.tab <- otu_table(subphylo.rel)
  MRA <- rowMeans(ra.tab)
  group <- tax_table(subphylo.rel)[,vect]
  mra.tab <- data.frame(group,MRA)
  colnames(mra.tab) <- c('level1', 'level2', 'MRA')
  #arrange the class table
  mra.tab_level1 = mra.tab %>% group_by(level1) %>% 
    summarise(sum_MRA = sum(MRA)) %>% 
    arrange(desc(sum_MRA))
  top_N_level1 = mra.tab_level1[1:N, ]$'level1'
  top_N_tab = mra.tab[mra.tab$'level1' %in% top_N_level1, ]
  mra.tab_level2 = top_N_tab %>% group_by(level2) %>% 
    summarise(sum_MRA = sum(MRA)) %>% 
    arrange(desc(sum_MRA))
  order_level2 = mra.tab_level2$'level2'
  top_N_tab$'level1' = factor(top_N_tab$'level1', ordered = T, levels = top_N_level1)
  top_N_tab$'level2' = factor(top_N_tab$'level2', ordered = T, levels = rev(order_level2))
  top_N_tab$MRA = top_N_tab$MRA*100
  return(top_N_tab)
}
```

Determine the relative abundance of the taxa at the phylum level.
```{r}
top10family <- arrange.tab(phylo, 10, 'Family', c(5,7))
mra.tab_level1 = top10family %>% group_by(level1) %>% 
  summarise(sum_MRA = sum(MRA)) %>% 
  arrange(desc(sum_MRA)) %>%
  dplyr::rename(Family = level1, prop = sum_MRA) %>%
  bind_rows(summarise_all(., ~if(is.numeric(.)) 100-sum(.) else "Others"))
mra.tab_level1
```
## Extract the phylogenetic trees of the top 5 familys
```{r}
fam_names <- c("KD4-96", "Defluviicoccaceae", "67-14", "Nitrosomonadaceae", "Comamonadaceae")
kd4_96_phylo <- subset_taxa(phylo, Family == "KD4-96")
kd4_96_tree <- phy_tree(kd4_96_phylo)
Deflu_phylo <- subset_taxa(phylo, Family == "Defluviicoccaceae")
Deflu_tree <- phy_tree(Deflu_phylo)
fam67_14_phylo <- subset_taxa(phylo, Family == "67-14")
fam67_14_tree <- phy_tree(fam67_14_phylo)
Nitro_phylo <- subset_taxa(phylo, Family == "Nitrosomonadaceae")
Nitro_tree <- phy_tree(Nitro_phylo)
Comamo_phylo <- subset_taxa(phylo, Family == "Comamonadaceae")
Comamo_tree <- phy_tree(Comamo_phylo)
# plot the trees
library(ape)
library(ggplot2)
library("phytools") # for sims and ASRs
library("ggtree")
g1 <- ggtree(kd4_96_tree, layout = 'circular', open.angle = 20)
g2 <- ggtree(Deflu_tree, layout = 'circular', open.angle = 20)
g3 <- ggtree(fam67_14_tree, layout = 'circular', open.angle = 20)
g4 <- ggtree(Nitro_tree, layout = 'circular', open.angle = 20)
g5 <- ggtree(Comamo_tree, layout = 'circular', open.angle = 20)
trees_plot <- cowplot::plot_grid(g1, g2, g3, g4, g5, nrow = 2)
# ggsave(file.path(save.dir.revision, "./trees_plot.pdf"), trees_plot, width = 6, height = 4.5, units = "in")
trees_plot
```
## Determine the phylogenitic distance between clades within each site among three layers
```{r}
# here we use mean pairwise distance to measure the phylogenetic distance between clades within samples
pd_kd4_96 <- mpd(t(data.frame(otu_table(kd4_96_phylo))), cophenetic(kd4_96_tree), abundance.weighted = F)
pd_Deflu <- mpd(t(data.frame(otu_table(Deflu_phylo))), cophenetic(Deflu_tree), abundance.weighted = F) 
pd_67_14 <- mpd(t(data.frame(otu_table(fam67_14_phylo))), cophenetic(fam67_14_tree), abundance.weighted = F) 
pd_Nitro <- mpd(t(data.frame(otu_table(Nitro_phylo))), cophenetic(Nitro_tree), abundance.weighted = F) 
pd_Comamo <- mpd(t(data.frame(otu_table(Comamo_phylo))), cophenetic(Comamo_tree), abundance.weighted = F) 

pd_top5 <- data.frame(Layers = metadata$Layer, kd4_96 = pd_kd4_96, Deflu = pd_Deflu, fam67_14 = pd_67_14, Nitro = pd_Nitro, Comamo = pd_Comamo)

# Box plot for evenness index using ggplot2
rep_str = c("kd4_96" = "KD4-96",
            "Deflu" = "Defluviicoccaceae",
            "fam67_14" = "67-14",
            "Nitro" = "Nitrosomonadaceae",
            "Comamo" = "Comamonadaceae")
# write a function to change the facet labels
facet_labeller <- function(variable,value){
  return(rep_str[value])
}

library(ggpubr)
library(ggplot2)
my_comparisons <- list( c("SUR", "SUB"), c("SUB", "PL"), c("SUR", "PL"))
MPD_top5 <- pd_top5 %>%
  pivot_longer(-c(Layers), names_to = "Family", values_to = "MPD") %>%
  mutate(Layers = factor(Layers, levels = c("SUR", "SUB", "PL"))) %>%
  mutate(Family = factor(Family, levels = c("kd4_96", "Deflu", "fam67_14", "Nitro", "Comamo"))) %>%
  ggplot(aes(x = Layers, y = MPD)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  facet_wrap(. ~ Family, scales = 'free', labeller = facet_labeller) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons, paired = TRUE, 
                     p.adjust.method = "BH", label = "p.signif", 
                     na.rm = T, size = 2.5, tip.length = 0.00, method = "wilcox.test") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'Mean pairwise distance (MPD)', fill='Layers') +
  main_theme + theme(legend.position = "none")
# save plot
# ggsave(file.path(save.dir.revision, "./MPD.pdf"), MPD_top5, width = 6, height = 4.5, units = "in")
MPD_top5
```

## Explore the charecteristics of MAGs
```{r}
MAG_df <- read.table(file = file.path(wd, 'revision/data/Widb.csv'),
                           sep = ',',  header = T, stringsAsFactors = F)
seq_df <- read.table(file = file.path(wd, 'revision/data/cleandata.txt'),
                           sep = '\t',  header = T, stringsAsFactors = F)
MAG_tab <- rbind(MAG_df %>% select(Layer, completeness, contamination) %>%
  pivot_longer(cols = -c("Layer"), names_to = "Index", values_to = "Value"),
  seq_df %>% select(Layer, Clean_data, Recovery_rate_MAGs) %>%
  pivot_longer(cols = -c("Layer"), names_to = "Index", values_to = "Value"))

MAG_tab %>% group_by(Layer, Index) %>%
  dplyr::summarise_all(list(mean = mean, sd = sd, se = ~sd(./sqrt(.))))
```

```{r}
# Box plot for evenness index using ggplot2
rep_str = c("Clean_data" = "Sequencing depth (G)",
            "completeness" = "Completeness (%)",
            "contamination" = "Contamination (%)",
            "Recovery_rate_MAGs" = "Recovery rate of MAGs")
# write a function to change the facet labels
facet_labeller <- function(variable,value){
  return(rep_str[value])
}
#ggplot
library(ggpubr)
library(ggplot2)
my_comparisons <- list( c("SUR", "SUB"), c("SUB", "PL"), c("SUR", "PL"))
MAG_plot <- MAG_tab %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL"))) %>%
  # mutate(Family = factor(Family, levels = c("kd4_96", "Deflu", "fam67_14", "Nitro", "Comamo"))) %>%
  ggplot(aes(x = Layer, y = Value)) + 
  geom_boxplot(width = 0.5, aes(fill = Layer)) +
  facet_wrap(. ~ Index, scales = 'free', labeller = facet_labeller) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons, paired = F, 
                     p.adjust.method = "BH", label = "p.signif", 
                     na.rm = T, size = 2.5, tip.length = 0.00, method = "wilcox.test") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layer', fill='Layer') +
  main_theme + theme(legend.position = "none")

# save plot
# ggsave(file.path(save.dir.revision, "./MAG_plot.pdf"), MAG_plot, width = 4.5, height = 4.5, units = "in")
MAG_plot
```
## All
```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
MAGs_all <- merge(dat3, tax_tab, by = "ID", all = T) # dat3, tax_tab在Metagenome_analysis.Rmd中
MAGs_all_df <- rbind(MAGs_all %>% filter(Phylum == "p__Acidobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Acidobacteriota", nrow(.))), 
  MAGs_all %>% filter(Phylum == "p__Chloroflexota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Chloroflexota", nrow(.))), 
  MAGs_all %>% filter(Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))),
  MAGs_all %>% filter(Class == "c__Gammaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(Class == "c__Alphaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Alphaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(Phylum == "p__Desulfobacterota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Desulfobacterota", nrow(.))))

library(tidytext)
all_num.mags.plot <- MAGs_all_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black', size = 9),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
# save plot
# ggsave(file.path(save.dir.revision, "./all_num.mags1.pdf"), all_num.mags.plot, width = 14.9, height = 11.9, units = "in")
all_num.mags.plot
```
## SUR
```{r, fig.align='center', fig.width=8, fig.height=4}
MAGs_all <- merge(dat3, tax_tab, by = "ID", all = T)

MAGs_SUR_df <- rbind(MAGs_all %>% filter(ID %in% grep("_SUR_", ID, value = T),
                    Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_SUR_", ID, value = T),
                    Class == "c__Alphaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))))

library(tidytext)
SUR_num.mags.plot <- MAGs_SUR_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
SUR_num.mags.plot
```

## SUB
```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
MAGs_SUB_df <- rbind(MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Phylum == "p__Acidobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Acidobacteriota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Phylum == "p__Chloroflexota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Chloroflexota", nrow(.))), 
   MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))),
  MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Class == "c__Gammaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Phylum == "p__Desulfobacterota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Desulfobacterota", nrow(.))))

library(tidytext)
SUB_num.mags.plot <- MAGs_SUB_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
SUB_num.mags.plot
```

## PF
```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
MAGs_PL_df <- rbind(MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Acidobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Acidobacteriota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Chloroflexota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Chloroflexota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))),
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Class == "c__Gammaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Desulfobacterota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Desulfobacterota", nrow(.))))

library(tidytext)
PL_num.mags.plot <- MAGs_PL_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
PL_num.mags.plot
```
# determine the difference in Eh among soil layers
```{r}
Eh_stat_data <- meta_dat %>% as.tibble() %>% select(Layer, Eh) %>%
   group_by(Layer) %>%
   summarise(mean.value = mean(Eh),
             sd.value = sd(Eh), count = n(),
             se.mean = sd.value/sqrt(count)) %>%
   mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL")))

Eh_paired_wilcox_test <- ggplot(Eh_stat_data, aes(x = Layer, y = mean.value, fill = Layer)) + 
   geom_bar(stat = "identity", color = "black", width = 0.7)+  
   geom_errorbar(aes(ymin = mean.value - se.mean, ymax = mean.value + se.mean), 
                 width = 0.2) +
   scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) + 
   xlab("Layers") + ylab("Values") + 
   scale_y_continuous(expand = expansion(mult = c(0, 0.35))) +
   main_theme

# save plot
# ggsave(file.path(save.dir.revision, "Eh_paired_wilcox_test.pdf"),
#        Eh_paired_wilcox_test, width = 2.5, height = 2.5, units = "in")
Eh_paired_wilcox_test
```

```{r}
fermen_pl_MAGs
MAGs_all %>% as.tibble() %>%
  filter(Category == "Fermentation", ID %in% grep("_PL_", ID, value = T)) %>%
  select(c(3, 4, 6, 8)) %>%
  group_by(Phylum, Order, Function) %>% 
  summarise(across(everything(), sum)) %>%
  write.table(file = file.path(save.dir.revision, "fermentation_pl_MAGs.csv"),
              row.names = F, sep = ",", quote = F)
 
knitr::kable(arrange(fermen_pl_MAGs, desc(presence_or_absent)))
```

```{r, fig.align='center', fig.height=8.9, fig.width=8.9}
# determine the gene presence in each genome
library(tidyverse)
library(stringr)

MAGs_name <- list.files("E:/permafrost/revision/data/KEGG_identifier_result/", 
           pattern = "*.result.txt")
MAGs_names <-  sapply(stringr::str_split(MAGs_name, ".result.txt"), `[`, 1)

temp_file <- list.files("E:/permafrost/revision/data/KEGG_identifier_result/", 
           pattern = "*.result.txt", full.names = T) %>%
  lapply(function(x) read.table(x, sep = "\t")) %>%
  Reduce(function(x, y) {merge(x, y, all = T, by = "V1")}, .) %>%
  `colnames<-`(c("KO", MAGs_names))

# write.table(temp_file, file.path(save.dir.revision, "genes_presence_MAGs.txt"), 
#               row.names = F, sep = "\t", na = "")

sel_ko_fermen_MAGs <- temp_file %>%
  right_join(sel_ko, by = "KO") %>%
  filter(Category == "Fermentation") %>%
  select(-c("Category", "Subcategory", "KO")) %>%
  mutate(across(where(is.numeric), replace_na, 0)) %>%
  group_by(Enzyme_protein_encoded) %>%
  summarise(across(everything(), sum)) %>%
  pivot_longer(cols = -c("Enzyme_protein_encoded"),
               names_to = "ID",
               values_to = "presence") %>%
  left_join(tax_tab, by = 'ID') %>%
  mutate(Layer = sapply(stringr::str_split(ID, "_"), `[`, 2)) %>%
  select(c("Layer", "Enzyme_protein_encoded", "presence", "Phylum")) %>%
  group_by(Layer, Phylum, Enzyme_protein_encoded) %>%
  summarise(across(everything(), sum)) %>%
  as.data.frame()


mycols <- c("#89c5da", "#ffc15c", "#74d944", "#CE50CA", "#5e738f", "#C0717C", "#CBD5ec", "#5F7FC7", 
                     "#00718b", "#d3d93e", "#169e77", "#d95f02", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD", 
                     "#D14285", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F", "#D1A33D", 
                     "#8A7C64", "#599861")


#plotting
fermen_genes_barplot <- sel_ko_fermen_MAGs %>%
    mutate(Phylum = recode(Phylum, p__Firmicutes_A = 'p__Firmicutes', 
                           p__Firmicutes_B = 'p__Firmicutes', 
                           p__Firmicutes_C = 'p__Firmicutes',
                           p__Desulfobacterota_E = "p__Desulfobacterota")) %>% 
  mutate(Phylum = sapply(stringr::str_split(Phylum, "__"), `[`, 2)) %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL"))) %>%
  ggplot(aes(fill = Phylum, y = presence, x = Enzyme_protein_encoded)) + 
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_manual(values = mycols) + 
  labs(x = 'Fermentation pathways', y = 'Number of genes') +
  scale_y_continuous(expand = expansion(mult = c(0, 0.15))) +
  facet_wrap(. ~ Layer, scales = "free_y", ncol = 1) +
  theme_bw()+
  theme(#legend.position = "none",
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour='black',size=12),
        axis.text.x = element_text(colour='black', size = 12, angle = 45, hjust = 1),
        legend.text=element_text(size=12),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "white"))
# save plot
ggsave(file.path(save.dir.revision, "fermen_genes_barplot.pdf"),
       fermen_genes_barplot, width = 8, height = 8.9, units = "in")
fermen_genes_barplot
```

```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
sel_ko_Nitro_MAGs <- temp_file %>%
  right_join(sel_ko, by = "KO") %>%
  filter(Category == "Nitrogen") %>%
  select(-c("Enzyme_protein_encoded", "Category", "KO")) %>%
  mutate(across(where(is.numeric), replace_na, 0)) %>%
  group_by(Subcategory) %>%
  summarise(across(everything(), sum)) %>%
  pivot_longer(cols = -c("Subcategory"),
               names_to = "ID",
               values_to = "presence") %>%
  left_join(tax_tab, by = 'ID') %>%
  mutate(Layer = sapply(stringr::str_split(ID, "_"), `[`, 2)) %>%
  select(c("Layer", "Subcategory", "presence", "Phylum")) %>%
  group_by(Layer, Phylum, Subcategory) %>%
  summarise(across(everything(), sum))


mycols <- c("#89c5da", "#ffc15c", "#74d944", "#CE50CA", "#5e738f", "#C0717C", "#CBD5ec", "#5F7FC7", 
                     "#00718b", "#d3d93e", "#169e77", "#d95f02", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD", 
                     "#D14285", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F", "#D1A33D", 
                     "#8A7C64", "#599861")

Nitro_genes_barplot <- sel_ko_fermen_MAGs %>%
    mutate(Phylum = recode(Phylum, p__Firmicutes_A = 'p__Firmicutes', 
                           p__Firmicutes_B = 'p__Firmicutes', 
                           p__Firmicutes_C = 'p__Firmicutes',
                           p__Desulfobacterota_E = "p__Desulfobacterota")) %>% 
  mutate(Phylum = sapply(stringr::str_split(Phylum, "__"), `[`, 2)) %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL"))) %>%
  ggplot(aes(fill = Phylum, y = presence, x = Subcategory)) + 
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_manual(values = mycols) + 
  labs(x = 'Nitrogen pathways', y = 'Number of genes') +
  scale_y_continuous(expand = expansion(mult = c(0, 0.15))) +
  facet_wrap(. ~ Layer, scales = "free_y", ncol = 1) +
  theme_bw()+
  theme(#legend.position = "none",
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour='black',size=12),
        axis.text.x = element_text(colour='black', size = 12, angle = 45, hjust = 1),
        legend.text=element_text(size=12),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "white"))
# save plot
# ggsave(file.path(save.dir.revision, "fermen_genes_barplot.pdf"),
#        fermen_genes_barplot, width = 8.9, height = 8.9, units = "in")
Nitro_genes_barplot
```

```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
sel_ko_Sulf_MAGs <- temp_file %>%
  right_join(sel_ko, by = "KO") %>%
  filter(Category == "Sulfur") %>%
  select(-c("Enzyme_protein_encoded", "Category", "KO")) %>%
  mutate(across(where(is.numeric), replace_na, 0)) %>%
  group_by(Subcategory) %>%
  summarise(across(everything(), sum)) %>%
  pivot_longer(cols = -c("Subcategory"),
               names_to = "ID",
               values_to = "presence") %>%
  left_join(tax_tab, by = 'ID') %>%
  mutate(Layer = sapply(stringr::str_split(ID, "_"), `[`, 2)) %>%
  select(c("Layer", "Subcategory", "presence", "Phylum")) %>%
  group_by(Layer, Phylum, Subcategory) %>%
  summarise(across(everything(), sum))


mycols <- c("#89c5da", "#ffc15c", "#74d944", "#CE50CA", "#5e738f", "#C0717C", "#CBD5ec", "#5F7FC7", 
                     "#00718b", "#d3d93e", "#169e77", "#d95f02", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD", 
                     "#D14285", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F", "#D1A33D", 
                     "#8A7C64", "#599861")

Sulf_genes_barplot <- sel_ko_fermen_MAGs %>%
    mutate(Phylum = recode(Phylum, p__Firmicutes_A = 'p__Firmicutes', 
                           p__Firmicutes_B = 'p__Firmicutes', 
                           p__Firmicutes_C = 'p__Firmicutes',
                           p__Desulfobacterota_E = "p__Desulfobacterota")) %>% 
  mutate(Phylum = sapply(stringr::str_split(Phylum, "__"), `[`, 2)) %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL"))) %>%
  ggplot(aes(fill = Phylum, y = presence, x = Subcategory)) + 
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_manual(values = mycols) + 
  labs(x = 'Sulful pathways', y = 'Number of genes') +
  scale_y_continuous(expand = expansion(mult = c(0, 0.15))) +
  facet_wrap(. ~ Layer, scales = "free_y", ncol = 1) +
  theme_bw()+
  theme(#legend.position = "none",
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour='black',size=12),
        axis.text.x = element_text(colour='black', size = 12, angle = 45, hjust = 1),
        legend.text=element_text(size=12),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "white"))
# save plot
# ggsave(file.path(save.dir.revision, "fermen_genes_barplot.pdf"),
#        fermen_genes_barplot, width = 8.9, height = 8.9, units = "in")
Sulf_genes_barplot
```


```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
sel_ko_other_MAGs <- temp_file %>%
  right_join(sel_ko, by = "KO") %>%
  filter(Category %in% c("Iron cycling", "As cycling", "Selenate reduction")) %>%
  select(-c("Enzyme_protein_encoded", "Category", "KO")) %>%
  mutate(across(where(is.numeric), replace_na, 0)) %>%
  group_by(Subcategory) %>%
  summarise(across(everything(), sum)) %>%
  pivot_longer(cols = -c("Subcategory"),
               names_to = "ID",
               values_to = "presence") %>%
  left_join(tax_tab, by = 'ID') %>%
  mutate(Layer = sapply(stringr::str_split(ID, "_"), `[`, 2)) %>%
  select(c("Layer", "Subcategory", "presence", "Phylum")) %>%
  group_by(Layer, Phylum, Subcategory) %>%
  summarise(across(everything(), sum))


mycols <- c("#89c5da", "#ffc15c", "#74d944", "#CE50CA", "#5e738f", "#C0717C", "#CBD5ec", "#5F7FC7", 
                     "#00718b", "#d3d93e", "#169e77", "#d95f02", "#D7C1B1", "#689030", "#AD6F3B", "#CD9BCD", 
                     "#D14285", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F", "#D1A33D", 
                     "#8A7C64", "#599861")

other_genes_barplot <- sel_ko_fermen_MAGs %>%
    mutate(Phylum = recode(Phylum, p__Firmicutes_A = 'p__Firmicutes', 
                           p__Firmicutes_B = 'p__Firmicutes', 
                           p__Firmicutes_C = 'p__Firmicutes',
                           p__Desulfobacterota_E = "p__Desulfobacterota")) %>% 
  mutate(Phylum = sapply(stringr::str_split(Phylum, "__"), `[`, 2)) %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL"))) %>%
  ggplot(aes(fill = Phylum, y = presence, x = Subcategory)) + 
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_manual(values = mycols) + 
  labs(x = 'Other pathways', y = 'Number of genes') +
  scale_y_continuous(expand = expansion(mult = c(0, 0.15))) +
  facet_wrap(. ~ Layer, scales = "free_y", ncol = 1) +
  theme_bw()+
  theme(#legend.position = "none",
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour='black',size=12),
        axis.text.x = element_text(colour='black', size = 12, angle = 45, hjust = 1),
        legend.text=element_text(size=12),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "white"))
# save plot
# ggsave(file.path(save.dir.revision, "fermen_genes_barplot.pdf"),
#        fermen_genes_barplot, width = 8.9, height = 8.9, units = "in")
other_genes_barplot
```

# explore the relationship between the dayas of the freeze thaw cycles and soil depth using TP data

```{r}
FTC_df <- read.table(file = file.path(wd, 'revision/data/FTC_data.csv'),
                           sep = ',',  header = T, stringsAsFactors = F)
# Fitting exponential decay curve
library(stats)
model <- nls(Days ~ a * exp(b * Depth) + c, data = FTC_df, start = list(a = 200, b = -0.25, c = 6))

#plot
FTC_plot <- ggplot(FTC_df, aes(x = Depth, y = Days)) + 
  geom_point(size = 3.5, alpha = 0.8, colour = "#D14285") +
  geom_smooth(method = "nls", 
              formula = "y ~ A * exp(B * x) + C", 
              method.args = list(start = c(A = 200, B = -0.25, C = 6)),
              size = 1, se = F, linetype = 1,  colour = 'black') +
  xlab("Depth (cm)") + ylab("Thawing time (Day)") +
  theme_bw() +
  theme(panel.grid = element_blank(), 
        axis.title = element_text(color = "black", size = 10),
        axis.ticks.length = unit(0.2, "lines"), 
        axis.ticks = element_line(color = "black"), 
        axis.line = element_blank(), 
        axis.text.y = element_text(colour = "black", size = 8), 
        axis.text.x = element_text(colour = "black", size = 8), 
        strip.text = element_text(size = 14), legend.position = "none")
# ggsave(file.path(save.dir.revision, "FTC_curve.pdf"),
#        FTC_plot, width = 4, height = 3.5, units = "in")
FTC_plot
```
# explore the relationships between the microbial processes and biogeochemical processes
```{r, fig.align='center', fig.height=8.9, fig.width=12.9}
# ch4_df <- read.table(file = file.path(wd, 'revision/data/CH4.csv'),
#                            sep = ',',  header = T, stringsAsFactors = F) %>%
#   mutate(Layer = sapply(stringr::str_split(Sample_id, "_"), '[', 1)) %>%
#   mutate(Site = sapply(stringr::str_split(Sample_id, "_"), '[', 2)) %>%
#   select(Layer, Site, CH4) %>%
#   group_by(Layer, Site) %>%
#   summarise(across(everything(), mean)) %>%
#   write.table(file = file.path(wd, 'revision/data/CH4_flux.csv'),
#                            sep = ',',  row.names = F)

# N_df <- read.table(file = file.path(wd, 'revision/data/N_processes.csv'),
#                            sep = ',',  header = T, stringsAsFactors = F) %>%
#   mutate(Site = sapply(stringr::str_split(Sample, "-"), '[', 1)) %>%
#   select(-c("Sample")) %>%
#   group_by(Layer, Site) %>%
#   summarise(across(everything(), mean)) %>%
#   write.table(file = file.path(wd, 'revision/data/N_processes_avg.csv'),
#                            sep = ',',  row.names = F)

# C_N_path2 %>% t() %>% as.data.frame() %>% select(C_names) %>%
#   rownames_to_column(var = "Sample") %>% 
#   write.table(file = file.path(wd, 'revision/data/C_path2.csv'),
#                             sep = ',',  row.names = F)

source("https://raw.githubusercontent.com/kangluyao/source_R/main/multivariables_regression_plot.R")

co2_df <- read.table(file = file.path(wd, 'revision/data/CO2_flux.csv'),
                           sep = ',',  header = T, stringsAsFactors = F) %>%
  select(-c("Site", "Cumulative.CO2.flux_5", "Cumulative.CO2.flux_15", "SOC")) 

# plot
carbon_process_names <- list(
  'Alpha.amylase' = "Alpha-amylase",
  'Glucoamylase' = "Glucoamylase",
  'Pullulanase' = "Pullulanase",
  'Isopullulanase' = "Isopullulanase",
  'Arabinofuranosidase' = "Arabinofuranosidase",
  'Beta_mannanase' = "Beta-mannanase",
  'Xylanase' = "Xylanase",
  'Xylose.isomerase' = "Xylose isomerase",
  'Beta.glucosidase' = "Beta-glucosidase",
  'Endoglucanase' = "Endoglucanase",
  'Exoglucanase' = "Exoglucanase",
  'Acetylglucosaminidase' = "Acetylglucosaminidase",
  'Endochitinase' = "Endochitinase",
  'Exochitinase' = "Exochitinase",
  'Pectinase' = "Pectinase",
  'Aryl.aldehyde.oxidase' = "Aryl-aldehyde oxidase",
  'Isocitrate.lyase' = "Isocitrate lyase",
  'Limonene.1..2.epoxide.hydrolase' = "Limonene-1, 2-epoxide hydrolase",
  'Malate.synthase' = "Malate synthase",
  'Vanillate.demethylase' = "Vanillate demethylase",
  'Glyoxal.oxidase' = "Glyoxal oxidase",
  'Phenol.oxidase..tyrosinase.' = "Phenol oxidase (tyrosinase)"
)

# write a function to change the facet labels
facet_labeller <- function(variable, value){
  return(carbon_process_names[value])
}

CO2_flux_plot <- co2_df %>%
  pivot_longer(cols = -c(Cumulative_CO2_flux_5_soil), names_to = "Crabon_processes", values_to = "value") %>% 
  mutate(Crabon_processes = factor(Crabon_processes, levels = colnames(co2_df)[-1])) %>%
  ggplot(aes(x = value, y = Cumulative_CO2_flux_5_soil)) +
  geom_point(shape = 19, colour = "#D14285")+
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(value, Cumulative_CO2_flux_5_soil, 
                       label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.05, label.y.npc = 0.95, size = 3.5) +
  facet_wrap( .~ Crabon_processes, scales = "free", ncol = 6, labeller = facet_labeller) +
  xlab('Gene abundance (TPM)') +
  ylab(expression(paste("Cumulative", "CO"[2], "-C", " release", " (mg CO"[2], "-C", " g"^-1, ")"))) +
  theme_bw() +
  theme(axis.title = element_text(size = 8), 
        axis.text = element_text(size = 7, colour = 'black'), 
        panel.grid = element_blank(),
        strip.text = element_text(size = 8))
ggsave(file.path(save.dir.revision, "CO2_flux_plot.pdf"),
       CO2_flux_plot, width = 13, height = 8, units = "in")
CO2_flux_plot
```


```{r, fig.align='center', fig.height=3.5, fig.width=8.9}
ch4_df <- read.table(file = file.path(wd, 'revision/data/CH4_flux.csv'),
                           sep = ',',  header = T, stringsAsFactors = F) %>%
  select(-c("Layer", "Site"))

ch4_process_names <- list(
  'Methyl.coenzyme.M.reductase' = "Methyl coenzyme M reductase",
  'Particulate.methane.monooxygenase' = "Particulate methane monooxygenase",
  'Soluable.methane.monooxygenase' = "Soluble methane monooxygenase"
)

# write a function to change the facet labels
facet_labeller <- function(variable, value){
  return(ch4_process_names[value])
}
ch4_flux_plot <- ch4_df %>% 
  pivot_longer(cols = -c(CH4), names_to = "ch4_processes", values_to = "value") %>% 
  mutate(ch4_processes = factor(ch4_processes, levels = colnames(ch4_df)[-1])) %>%
  ggplot(aes(x = value, y = CH4)) +
  geom_point(shape = 19, colour = "#D14285")+
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(value, CH4, 
                       label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.25, label.y.npc = 0.95, size = 3.5) +
  facet_wrap( .~ ch4_processes, scales = "free", ncol = 3, labeller = facet_labeller) +
  xlab('Gene abundance (TPM)') +
  ylab(expression(paste("CH"[4], " production", " (ng", " g"^-1, " day"^-1, ")"))) +
  theme_bw() +
  theme(axis.title = element_text(size = 8), 
        axis.text = element_text(size = 7, colour = 'black'), 
        panel.grid = element_blank(),
        strip.text = element_text(size = 8))
ggsave(file.path(save.dir.revision, "ch4_flux_plot.pdf"),
       ch4_flux_plot, width = 6.9, height = 2.5, units = "in")
ch4_flux_plot
```


```{r, fig.align='center', fig.height=5.5, fig.width=8.9}
# C_N_path2[N_names, ] %>% t() %>%
#   as.data.frame() %>%
#   rownames_to_column(var = "Sample") %>%
#   write.table(file = file.path(wd, 'revision/data/N_path2.csv'),
#                             sep = ',',  row.names = F)

N_df <- read.table(file = file.path(wd, 'revision/data/N_processes_avg.csv'),
                           sep = ',',  header = T, stringsAsFactors = F) %>%
  select(-c("Layer", "Site")) %>%
  as.data.frame()

N_proces_var <- c("Gross.N.mineralization", "Microbial.immobilization", "Nitrification")
N_vars <- c("gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC", "nmo",
            "glnA",
            "amoA", "amoB",  "amoC", "hao")

Gross.N.mineralization_plot <- N_df %>% 
  select("Gross.N.mineralization", "gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC") %>%
  pivot_longer(cols = -c(Gross.N.mineralization), names_to = "processes", values_to = "value") %>% 
  mutate(N.mineralization_processes = factor(N.mineralization_processes, 
                                             levels = c("gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC"))) %>%
  ggplot(aes(x = value, y = Gross.N.mineralization)) +
  geom_point(shape = 19, colour = "#D14285")+
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(value, Gross.N.mineralization, 
                       label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.25, label.y.npc = 0.95, size = 3.5) +
  facet_wrap( .~ N.mineralization_processes, scales = "free", ncol = 3) +
  xlab('Gene abundance (TPM)') +
  ylab(expression(paste("Gross N mineralization", " (mg", " kg"^-1, " day"^-1, ")"))) +
  theme_bw() +
  theme(axis.title = element_text(size = 8), 
        axis.text = element_text(size = 7, colour = 'black'), 
        panel.grid = element_blank(),
        strip.text = element_text(size = 8))
# ggsave(file.path(save.dir.revision, "ch4_flux_plot.pdf"),
#        ch4_flux_plot, width = 6.9, height = 2.5, units = "in")
Gross.N.mineralization_plot
```

```{r, fig.align='center', fig.height=4.5, fig.width=5.5}
Nitrification_plot <- N_df %>% 
  select("Nitrification", "amoA", "amoB",  "amoC", "hao") %>%
  pivot_longer(cols = -c(Nitrification), names_to = "Nitrification_processes", values_to = "value") %>% 
  mutate(Nitrification_processes = factor(Nitrification_processes, 
                                             levels = c("amoA", "amoB",  "amoC", "hao"))) %>%
  ggplot(aes(x = value, y = Nitrification)) +
  geom_point(shape = 19, colour = "#D14285")+
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(value, Nitrification, 
                       label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.25, label.y.npc = 0.95, size = 3.5) +
  facet_wrap( .~ Nitrification_processes, scales = "free", ncol = 2) +
  xlab('Gene abundance (TPM)') +
  ylab(expression(paste("Nitrification", " (mg N", " kg"^-1, " day"^-1, ")"))) +
  theme_bw() +
  theme(axis.title = element_text(size = 8), 
        axis.text = element_text(size = 7, colour = 'black'), 
        panel.grid = element_blank(),
        strip.text = element_text(size = 8))
# ggsave(file.path(save.dir.revision, "ch4_flux_plot.pdf"),
#        ch4_flux_plot, width = 6.9, height = 2.5, units = "in")
Nitrification_plot
```

```{r, fig.align='center', fig.height=3.5, fig.width=3.5}
Microbial.immobilization_plot <- N_df %>% 
  select("Microbial.immobilization", "glnA") %>%
  ggplot(aes(x = glnA, y = Microbial.immobilization)) +
  geom_point(shape = 19, colour = "#D14285")+
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(glnA, Microbial.immobilization, 
                       label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.25, label.y.npc = 0.95, size = 3.5) +
  xlab('Gene abundance (TPM)') +
  ylab(expression(paste("Microbial immobilization", " (mg N", " kg"^-1, " day"^-1, ")"))) +
  theme_bw() +
  theme(axis.title = element_text(size = 8), 
        axis.text = element_text(size = 7, colour = 'black'), 
        panel.grid = element_blank(),
        strip.text = element_text(size = 8))
# ggsave(file.path(save.dir.revision, "ch4_flux_plot.pdf"),
#        ch4_flux_plot, width = 6.9, height = 2.5, units = "in")
Microbial.immobilization_plot
```


```{r, fig.align='center', fig.height=5.5, fig.width=8.9}
plot_N <- rbind(N_df %>% 
  select("Gross.N.mineralization", "gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC") %>%
  pivot_longer(cols = -c(Gross.N.mineralization), names_to = "genes", values_to = "value") %>%
  pivot_longer(cols = -c(genes, value), names_to = "processes", values_to = "value_1"), 
  N_df %>% 
  select("Nitrification", "amoA", "amoB",  "amoC", "hao") %>%
  pivot_longer(cols = -c(Nitrification), names_to = "genes", values_to = "value") %>%
  pivot_longer(cols = -c(genes, value), names_to = "processes", values_to = "value_1"), 
  N_df %>% 
  select("Microbial.immobilization", "glnA") %>%
  pivot_longer(cols = -c(Microbial.immobilization), names_to = "genes", values_to = "value") %>%
  pivot_longer(cols = -c(genes, value), names_to = "processes", values_to = "value_1"))

N_process_plot <- plot_N %>% 
  mutate(processes = factor(processes, 
                            levels = c("Gross.N.mineralization", "Microbial.immobilization", "Nitrification"))) %>%
  mutate(genes = factor(genes, levels = c(c("gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC", "nmo",
            "glnA", "amoA", "amoB",  "amoC", "hao")))) %>%
  ggplot(aes(x = value, y = value_1)) +
  geom_point(shape = 19, colour = "#D14285")+
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(value, value_1, 
                       label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.25, label.y.npc = 0.95, size = 3.5) +
  facet_wrap(genes ~ processes, scales = "free") +
  xlab('Gene abundance (TPM)') +
  ylab(expression(paste("Nitrogen processes", " (mg N", " kg"^-1, " day"^-1, ")"))) +
  theme_bw() +
  theme(axis.title = element_text(size = 8), 
        axis.text = element_text(size = 7, colour = 'black'), 
        panel.grid = element_blank(),
        strip.text = element_text(size = 8))
ggsave(file.path(save.dir.revision, "N_process_plot.pdf"),
       N_process_plot, width = 8.9, height = 6.9, units = "in")
N_process_plot
```












Input files
```{r}
sur_input_table <- file.path(wd_fun, "metabolic/METABOLIC_70_layer_rare/METABOLIC_70_sur/METABOLIC_Figures_Input/Metabolic_Sankey_diagram_input.txt")
sub_input_table <- file.path(wd_fun, "metabolic/METABOLIC_70_layer_rare/METABOLIC_70_sub/METABOLIC_Figures_Input/Metabolic_Sankey_diagram_input.txt")
pl_input_table <- file.path(wd_fun, "metabolic/METABOLIC_70_layer_rare/METABOLIC_70_pl/METABOLIC_Figures_Input/Metabolic_Sankey_diagram_input.txt")
```

Loading packages

```{r}
pacman::p_load(ggthemes, networkD3, data.table)
```

Generate a snakey plot with reactions as node, and edges are MAGs, and are colored by taxonomic group.

```{r}
# Load the energy flow diagram input ---------------------------
table_sur <- read.table(sur_input_table, header = F, sep = "\t")
colnames(table_sur) <- c("Taxa", "Reaction", "Freq")
table_sub <- read.table(sub_input_table, header = F, sep = "\t")
colnames(table_sub) <- c("Taxa", "Reaction", "Freq")
table_pl <- read.table(pl_input_table, header = F, sep = "\t")
colnames(table_pl) <- c("Taxa", "Reaction", "Freq")
```

```{r}
# Rename the categories ---------------------------
table_sur$Category <- ifelse(grepl("C-S", table_sur$Reaction), "Carbon", 
                               ifelse(grepl("N-S", table_sur$Reaction), "Nitrogen",
                                      ifelse(grepl("S-S", table_sur$Reaction), "Sulfur",
                                             ifelse(grepl("O-S", table_sur$Reaction), "Others",""))))
table_sur$Category <- as.factor(table_sur$Category)

table_sub$Category <- ifelse(grepl("C-S", table_sub$Reaction), "Carbon", 
                               ifelse(grepl("N-S", table_sub$Reaction), "Nitrogen",
                                      ifelse(grepl("S-S", table_sub$Reaction), "Sulfur",
                                             ifelse(grepl("O-S", table_sub$Reaction), "Others",""))))
table_sub$Category <- as.factor(table_sub$Category)

table_pl$Category <- ifelse(grepl("C-S", table_pl$Reaction), "Carbon", 
                               ifelse(grepl("N-S", table_pl$Reaction), "Nitrogen",
                                      ifelse(grepl("S-S", table_pl$Reaction), "Sulfur",
                                             ifelse(grepl("O-S", table_pl$Reaction), "Others",""))))
table_pl$Category <- as.factor(table_pl$Category)
```

```{r}
# reformate the data
table_all <- rbind(cbind(Layer = rep("SUR", nrow(table_sur)), table_sur),
                   cbind(Layer = rep("SUB", nrow(table_sub)), table_sub),
                   cbind(Layer = rep("PL", nrow(table_pl)), table_pl)) %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL")))

taxa_names <- table_all %>% select(Taxa, Freq) %>%
  group_by(Taxa) %>%
  summarise(sum = sum(Freq)) %>%
  arrange(desc(sum)) %>%
  pull(Taxa)
reaction_orders <- c(sort(grep('C-', unique(table_all$Reaction), value = T), decreasing = F),
                     sort(grep('N-', unique(table_all$Reaction), value = T), decreasing = F),
                     sort(grep('S-S', unique(table_all$Reaction), value = T), decreasing = F),
                     sort(grep('O-', unique(table_all$Reaction), value = T), decreasing = F))

table_all <- table_all %>%
  mutate(Taxa = factor(Taxa, levels = taxa_names)) %>%
  mutate(Category = factor(Category, levels = c('Carbon', 'Nitrogen', 
                                                'Sulfur', 'Others'))) %>%
  mutate(Reaction = factor(Reaction, levels = reaction_orders))

sankey <- data.table::rbindlist(list(table_all[c("Layer", "Taxa", "Freq")],
                         table_all[c("Taxa", "Reaction", "Freq")]))
names(sankey) <- c('source', 'target', 'value')
knitr::kable(sankey[1:5, ])
```
```{r}
# Make a connection data frame
links <- sankey 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes <- data.frame(
  name = c(as.character(links$source), as.character(links$target)) %>% 
    unique()
)
nodes <- data.frame(
  name = c("SUR", "SUB", "PL", taxa_names,
           sort(grep('C-', unique(table_all$Reaction), value = T), decreasing = F),
           sort(grep('N-', unique(table_all$Reaction), value = T), decreasing = F),
           sort(grep('S-S', unique(table_all$Reaction), value = T), decreasing = F),
           sort(grep('O-', unique(table_all$Reaction), value = T), decreasing = F))
)
# With networkD3, connection must be provided using id, not using real name like in the links dataframe.. So we need to reformat it.
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1
```

```{r}
# Set color of connections
# Add a 'group' column to each connection:
ass_link_col_fun <- function(df) {
  if (df["source"] == taxa_names[1] | df["target"] == taxa_names[1]) {
    group = 'a'
  } else if (df["source"] == taxa_names[2] | df["target"] == taxa_names[2]) {
    group = 'b'
  } else if (df["source"] == taxa_names[3] | df["target"] == taxa_names[3]) {
    group = 'c'
  } else if (df["source"] == taxa_names[4] | df["target"] == taxa_names[4]) {
    group = 'd'
  } else if (df["source"] == taxa_names[5] | df["target"] == taxa_names[5]) {
    group = 'e'
  } else if (df["source"] == taxa_names[6] | df["target"] == taxa_names[6]) {
    group = 'f'
  } else if (df["source"] == taxa_names[7] | df["target"] == taxa_names[7]) {
    group = 'g'
  } else if (df["source"] == taxa_names[8] | df["target"] == taxa_names[8]) {
    group = 'h'
  } else if (df["source"] == taxa_names[9] | df["target"] == taxa_names[9]) {
    group = 'i'
  } else if (df["source"] == taxa_names[10] | df["target"] == taxa_names[10]) {
    group = 'j'
  } else if (df["source"] == taxa_names[11] | df["target"] == taxa_names[11]) {
    group = 'k'
  } else if (df["source"] == taxa_names[12] | df["target"] == taxa_names[12]) {
    group = 'l'
  } else if (df["source"] == taxa_names[13] | df["target"] == taxa_names[13]) {
    group = 'm'
  } else if (df["source"] == taxa_names[14] | df["target"] == taxa_names[14]) {
    group = 'n'
  } else if (df["source"] == taxa_names[15] | df["target"] == taxa_names[15]) {
    group = 'o'
  } else if (df["source"] == taxa_names[16] | df["target"] == taxa_names[16]) {
    group = 'p'
  } else if (df["source"] == taxa_names[17] | df["target"] == taxa_names[17]) {
    group = 'q'
  } else if (df["source"] == taxa_names[18] | df["target"] == taxa_names[18]) {
    group = 'r'
  } else if (df["source"] == taxa_names[19] | df["target"] == taxa_names[19]) {
    group = 's'
  } else if (df["source"] == taxa_names[20] | df["target"] == taxa_names[20]) {
    group = 't'
  } else if (df["source"] == taxa_names[21] | df["target"] == taxa_names[21]) {
    group = 'u'
  } else if (df["source"] == taxa_names[21] | df["target"] == taxa_names[21]) {
    group = 'u'
  }
  return(group)
}
links$group <- apply(links, 1, ass_link_col_fun)

# Add a 'group' column to each node. Here I decide to put all of them in the same group to make them grey
nodes$group <- as.factor(c("my_unique_group"))
links[1:5, ]
nodes[1:5, ]
```

```{r}
# Give a color for each group
my_color <- 'd3.scaleOrdinal() .domain(["a", "b", "c", "d", "e", "f", 
"g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", 
"Carbon", "Nitrogen", "Sulfur", "Others", "my_unique_group"]) .range(["#4970a0", "#f8696b", "#FCCDE5",
"#6A3D9A", "#1F78B4", "#FDB462", "#80B1D3", "#BEBADA", "#c988d1", "#A6CEE3",
"#8DD3C7", "#B2DF8A", "#33A02C", "#FB9A99", "#FFFFB3", "#FDBF6F", "#E31A1C", "#d5416a", 
"#FB8072", "#d87a71","#6a75d5", "#76c2d7", "#d87a71", "#FF7F00", "#836834", 
"#c0c0c0"])'

# Make the Network. I call my colour scale with the colourScale argument
P <- sankeyNetwork(Links = links, Nodes = nodes, Source = "IDsource", Target = "IDtarget", 
                   Value = "value", NodeID = "name", colourScale = my_color, 
                   LinkGroup = "group", NodeGroup = "group", iterations = 0, 
                   fontFamily = 'Arial', fontSize = 15, 
                   nodeWidth = 20, nodePadding = 10, height = 700, width = 1600,
                   sinksRight = F)
# save the widget
# 推荐使用htmlwidgets
library(htmlwidgets)
# saveWidget(p, file = paste0(sankey.plots.folder, "/sankey_plot.html"))
# library(webshot)
# install phantom:
# webshot::install_phantomjs()
# Make a webshot in pdf : high quality but can not choose printed zone
# webshot(paste0(sankey.plots.folder, "/sankey_plot.html"), delay = 0.2)
P
```