---
title: "Analysis for revision"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set work and saving directory
```{r}
wd <- 'e:/permafrost'
setwd(wd)
save.dir.revision <- file.path(wd,"revision/results")
if (!dir.exists(save.dir.revision)) {
  dir.create(save.dir.revision)
}
```

## Reading data
```{r}
HF_df <- read.table(file = file.path(wd, 'revision/data/HF.csv'),
                           sep = ',',  header = T, stringsAsFactors = F)
```
Test the human disturbance on the Tibetan plateau
```{r}
# plot
library(gghalves)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
my_comparisons <- list(c('Sampling_sites', 'Cities'))
p1 <- HF_df %>%
  mutate(Type = factor(Type, levels = c("Sampling_sites", "Cities"))) %>%
  ggplot(aes(Type, HF_2000_2015, fill = Type)) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Type), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_x_discrete(labels = c("Sampling_sites" = "Sampling sites", "Cities" = "Cities")) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = "Human footprint index") +
  scale_fill_manual(values = c("#f8766d", "#a3a500")) +
  main_theme + theme(legend.position = "none")
# save plot
# ggsave(file.path(save.dir.revision, "./HF.pdf"), p1, width = 3.5, height = 2.5, units = "in")
p1
```

## Explor the phylogenecy of specific taxonomic groups and comparing the phylogenetic difference among layers
```{r}
# First, we want to know which taxonomic groups are dominant taxa in our samples


```

```{r}
# write a function for data organization
arrange.tab <- function(phylo, N, taxrank, vect) {
  subphylo <- tax_glom(phylo, taxrank)
  subphylo.rel <- microbiome::transform(subphylo, "compositional")
  ra.tab <- otu_table(subphylo.rel)
  MRA <- rowMeans(ra.tab)
  group <- tax_table(subphylo.rel)[,vect]
  mra.tab <- data.frame(group,MRA)
  colnames(mra.tab) <- c('level1', 'level2', 'MRA')
  #arrange the class table
  mra.tab_level1 = mra.tab %>% group_by(level1) %>% 
    summarise(sum_MRA = sum(MRA)) %>% 
    arrange(desc(sum_MRA))
  top_N_level1 = mra.tab_level1[1:N, ]$'level1'
  top_N_tab = mra.tab[mra.tab$'level1' %in% top_N_level1, ]
  mra.tab_level2 = top_N_tab %>% group_by(level2) %>% 
    summarise(sum_MRA = sum(MRA)) %>% 
    arrange(desc(sum_MRA))
  order_level2 = mra.tab_level2$'level2'
  top_N_tab$'level1' = factor(top_N_tab$'level1', ordered = T, levels = top_N_level1)
  top_N_tab$'level2' = factor(top_N_tab$'level2', ordered = T, levels = rev(order_level2))
  top_N_tab$MRA = top_N_tab$MRA*100
  return(top_N_tab)
}
```
Determine the relative abundance of the taxa at the phylum level.
```{r}
top10family <- arrange.tab(phylo, 10, 'Family', c(5,7))
mra.tab_level1 = top10family %>% group_by(level1) %>% 
  summarise(sum_MRA = sum(MRA)) %>% 
  arrange(desc(sum_MRA)) %>%
  dplyr::rename(Family = level1, prop = sum_MRA) %>%
  bind_rows(summarise_all(., ~if(is.numeric(.)) 100-sum(.) else "Others"))
mra.tab_level1
```
## Extract the phylogenetic trees of the top 5 familys
```{r}
fam_names <- c("KD4-96", "Defluviicoccaceae", "67-14", "Nitrosomonadaceae", "Comamonadaceae")
kd4_96_phylo <- subset_taxa(phylo, Family == "KD4-96")
kd4_96_tree <- phy_tree(kd4_96_phylo)
Deflu_phylo <- subset_taxa(phylo, Family == "Defluviicoccaceae")
Deflu_tree <- phy_tree(Deflu_phylo)
fam67_14_phylo <- subset_taxa(phylo, Family == "67-14")
fam67_14_tree <- phy_tree(fam67_14_phylo)
Nitro_phylo <- subset_taxa(phylo, Family == "Nitrosomonadaceae")
Nitro_tree <- phy_tree(Nitro_phylo)
Comamo_phylo <- subset_taxa(phylo, Family == "Comamonadaceae")
Comamo_tree <- phy_tree(Comamo_phylo)
# plot the trees
library(ape)
library(ggplot2)
library("phytools") # for sims and ASRs
library("ggtree")
g1 <- ggtree(kd4_96_tree, layout = 'circular', open.angle = 20)
g2 <- ggtree(Deflu_tree, layout = 'circular', open.angle = 20)
g3 <- ggtree(fam67_14_tree, layout = 'circular', open.angle = 20)
g4 <- ggtree(Nitro_tree, layout = 'circular', open.angle = 20)
g5 <- ggtree(Comamo_tree, layout = 'circular', open.angle = 20)
trees_plot <- cowplot::plot_grid(g1, g2, g3, g4, g5, nrow = 2)
# ggsave(file.path(save.dir.revision, "./trees_plot.pdf"), trees_plot, width = 6, height = 4.5, units = "in")
trees_plot
```
## Determine the phylogenitic distance between clades within each site among three layers
```{r}
# here we use mean pairwise distance to measure the phylogenetic distance between clades within samples
pd_kd4_96 <- mpd(t(data.frame(otu_table(kd4_96_phylo))), cophenetic(kd4_96_tree), abundance.weighted = F)
pd_Deflu <- mpd(t(data.frame(otu_table(Deflu_phylo))), cophenetic(Deflu_tree), abundance.weighted = F) 
pd_67_14 <- mpd(t(data.frame(otu_table(fam67_14_phylo))), cophenetic(fam67_14_tree), abundance.weighted = F) 
pd_Nitro <- mpd(t(data.frame(otu_table(Nitro_phylo))), cophenetic(Nitro_tree), abundance.weighted = F) 
pd_Comamo <- mpd(t(data.frame(otu_table(Comamo_phylo))), cophenetic(Comamo_tree), abundance.weighted = F) 

pd_top5 <- data.frame(Layers = metadata$Layer, kd4_96 = pd_kd4_96, Deflu = pd_Deflu, fam67_14 = pd_67_14, Nitro = pd_Nitro, Comamo = pd_Comamo)

# Box plot for evenness index using ggplot2
rep_str = c("kd4_96" = "KD4-96",
            "Deflu" = "Defluviicoccaceae",
            "fam67_14" = "67-14",
            "Nitro" = "Nitrosomonadaceae",
            "Comamo" = "Comamonadaceae")
# write a function to change the facet labels
facet_labeller <- function(variable,value){
  return(rep_str[value])
}

library(ggpubr)
library(ggplot2)
my_comparisons <- list( c("SUR", "SUB"), c("SUB", "PL"), c("SUR", "PL"))
MPD_top5 <- pd_top5 %>%
  pivot_longer(-c(Layers), names_to = "Family", values_to = "MPD") %>%
  mutate(Layers = factor(Layers, levels = c("SUR", "SUB", "PL"))) %>%
  mutate(Family = factor(Family, levels = c("kd4_96", "Deflu", "fam67_14", "Nitro", "Comamo"))) %>%
  ggplot(aes(x = Layers, y = MPD)) + 
  geom_boxplot(width = 0.5, aes(fill = Layers)) +
  facet_wrap(. ~ Family, scales = 'free', labeller = facet_labeller) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons, paired = TRUE, 
                     p.adjust.method = "BH", label = "p.signif", 
                     na.rm = T, size = 2.5, tip.length = 0.00, method = "wilcox.test") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layers', y = 'Mean pairwise distance (MPD)', fill='Layers') +
  main_theme + theme(legend.position = "none")
# save plot
# ggsave(file.path(save.dir.revision, "./MPD.pdf"), MPD_top5, width = 6, height = 4.5, units = "in")
MPD_top5
```

## Explore the charecteristics of MAGs
```{r}
MAG_df <- read.table(file = file.path(wd, 'revision/data/Widb.csv'),
                           sep = ',',  header = T, stringsAsFactors = F)
seq_df <- read.table(file = file.path(wd, 'revision/data/cleandata.txt'),
                           sep = '\t',  header = T, stringsAsFactors = F)
MAG_tab <- rbind(MAG_df %>% select(Layer, completeness, contamination) %>%
  pivot_longer(cols = -c("Layer"), names_to = "Index", values_to = "Value"),
  seq_df %>% select(Layer, Clean_data, Recovery_rate_MAGs) %>%
  pivot_longer(cols = -c("Layer"), names_to = "Index", values_to = "Value"))

# Box plot for evenness index using ggplot2
rep_str = c("Clean_data" = "Sequencing depth (G)",
            "completeness" = "Completeness (%)",
            "contamination" = "Contamination (%)",
            "Recovery_rate_MAGs" = "Recovery rate of MAGs")
# write a function to change the facet labels
facet_labeller <- function(variable,value){
  return(rep_str[value])
}
#ggplot
library(ggpubr)
library(ggplot2)
my_comparisons <- list( c("SUR", "SUB"), c("SUB", "PL"), c("SUR", "PL"))
MAG_plot <- MAG_tab %>%
  mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL"))) %>%
  # mutate(Family = factor(Family, levels = c("kd4_96", "Deflu", "fam67_14", "Nitro", "Comamo"))) %>%
  ggplot(aes(x = Layer, y = Value)) + 
  geom_boxplot(width = 0.5, aes(fill = Layer)) +
  facet_wrap(. ~ Index, scales = 'free', labeller = facet_labeller) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  stat_compare_means(comparisons = my_comparisons, paired = F, 
                     p.adjust.method = "BH", label = "p.signif", 
                     na.rm = T, size = 2.5, tip.length = 0.00, method = "wilcox.test") +
  #scale_fill_manual(values= cols)+
  labs(x = 'Layer', fill='Layer') +
  main_theme + theme(legend.position = "none")

# save plot
# ggsave(file.path(save.dir.revision, "./MAG_plot.pdf"), MAG_plot, width = 4.5, height = 4.5, units = "in")
MAG_plot
```
## All
```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
MAGs_all_df <- rbind(MAGs_all %>% filter(Phylum == "p__Acidobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Acidobacteriota", nrow(.))), 
  MAGs_all %>% filter(Phylum == "p__Chloroflexota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Chloroflexota", nrow(.))), 
  MAGs_all %>% filter(Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))),
  MAGs_all %>% filter(Class == "c__Gammaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(Class == "c__Alphaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Alphaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(Phylum == "p__Desulfobacterota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Desulfobacterota", nrow(.))))

library(tidytext)
all_num.mags.plot <- MAGs_all_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black', size = 9),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
# save plot
# ggsave(file.path(save.dir.revision, "./all_num.mags1.pdf"), all_num.mags.plot, width = 13.9, height = 8.9, units = "in")
all_num.mags.plot
```




## SUR
```{r, fig.align='center', fig.width=8, fig.height=4}
MAGs_all <- merge(dat3, tax_tab, by = "ID", all = T)

MAGs_SUR_df <- rbind(MAGs_all %>% filter(ID %in% grep("_SUR_", ID, value = T),
                    Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_SUR_", ID, value = T),
                    Class == "c__Alphaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))))

library(tidytext)
SUR_num.mags.plot <- MAGs_SUR_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
SUR_num.mags.plot
```

## SUB
```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
MAGs_SUB_df <- rbind(MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Phylum == "p__Acidobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Acidobacteriota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Phylum == "p__Chloroflexota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Chloroflexota", nrow(.))), 
   MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))),
  MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Class == "c__Gammaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_SUB_", ID, value = T),
                    Phylum == "p__Desulfobacterota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Desulfobacterota", nrow(.))))

library(tidytext)
SUB_num.mags.plot <- MAGs_SUB_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
SUB_num.mags.plot
```

## PF
```{r, fig.align='center', fig.width=12.9, fig.height=8.9}
MAGs_PL_df <- rbind(MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Acidobacteriota") %>% select(c(2, 3, 4)) %>%
                      group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  mutate(Taxa = rep("p__Acidobacteriota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Chloroflexota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Chloroflexota", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Actinobacteriota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Actinobacteriota", nrow(.))),
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Class == "c__Gammaproteobacteria") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("c__Gammaproteobacteria", nrow(.))), 
  MAGs_all %>% filter(ID %in% grep("_PL_", ID, value = T),
                    Phylum == "p__Desulfobacterota") %>% select(c(2, 3, 4)) %>%
    group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(Function %in% c(C_N_metapathway, other_metapathway), presence_or_absent >= 1) %>% 
    as.data.frame() %>%
  arrange(desc(presence_or_absent)) %>%
  mutate(Taxa = rep("p__Desulfobacterota", nrow(.))))

library(tidytext)
PL_num.mags.plot <- MAGs_PL_df %>%
  mutate(Taxa = as.factor(Taxa),
           Function = reorder_within(Function, presence_or_absent, Taxa)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Taxa)) +
  geom_bar(stat = "identity", width = 0.7, position = position_dodge(0.7)) +
  # scale_fill_manual(values = rep(my_col, 4)) +
  facet_wrap(. ~ Taxa, scales = 'free') +
  scale_x_reordered() +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.key.size = unit(1, "line"),
        legend.position = "none") +
  # guides(fill = guide_legend(ncol = 1, reverse = TRUE)) +
  coord_flip()
PL_num.mags.plot
```
# determine the difference in Eh among soil layers
```{r}
Eh_stat_data <- meta_dat %>% as.tibble() %>% select(Layer, Eh) %>%
   group_by(Layer) %>%
   summarise(mean.value = mean(Eh),
             sd.value = sd(Eh), count = n(),
             se.mean = sd.value/sqrt(count)) %>%
   mutate(Layer = factor(Layer, levels = c("SUR", "SUB", "PL")))

Eh_paired_wilcox_test <- ggplot(Eh_stat_data, aes(x = Layer, y = mean.value, fill = Layer)) + 
   geom_bar(stat = "identity", color = "black", width = 0.7)+  
   geom_errorbar(aes(ymin = mean.value - se.mean, ymax = mean.value + se.mean), 
                 width = 0.2) +
   scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) + 
   xlab("Layers") + ylab("Values") + 
   scale_y_continuous(expand = expansion(mult = c(0, 0.35))) +
   main_theme

# save plot
ggsave(file.path(save.dir.revision, "Eh_paired_wilcox_test.pdf"),
       Eh_paired_wilcox_test, width = 2.5, height = 2.5, units = "in")
Eh_paired_wilcox_test
```