---
title: "Species-occurence analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output: html_notebook
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set work directory
```{r}
setwd('e:/permafrost/')
wd_16s <- file.path(getwd(),"data/16S")
# if (!dir.exists(wd_16s)) {
#   dir.create(wd_16s)
# }
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```
Read data
```{r}
source("script/read_data.R")
```
Extract the cor taxa with the mean relative abundance greater than 0.01%
```{r}
phylo_rel <- transform_sample_counts(phylo, function(x) x / sum(x)) # transfor the counts table to relative abundance table 
phylo_rel_core <- phyloseq::filter_taxa(phylo_rel, function(x) mean(x) > .0001, TRUE)
ntaxa(phylo_rel_core)
```
```{r}
# 读取 OTU 丰度表和样本分组
otu_specif <- data.frame(otu_table(phylo_rel_core), Taxonomy = data.frame(tax_table(phylo_rel_core))$Phylum) 
# 计算各组样本中，OTU 的特异性(specificity)和占有率(occupancy)
Nindividuals_S <- rep(0, nrow(otu_specif))
for (i in unique(metadata$Layer)) {
  otu_group_i <- otu_specif[ ,rownames(subset(metadata, Layer == i))]
  Nindividuals_S <- Nindividuals_S + rowMeans(otu_group_i)  #计算 Nindividuals S
}
```
```{r}
spec_occu <- NULL
for (i in unique(metadata$Layer)) {
  otu_group_i <- otu_specif[ ,rownames(subset(metadata, Layer == i))]
  Nindividuals_SH <- apply(otu_group_i, 1, mean)  #计算 Nindividuals SH
  Specificity <- Nindividuals_SH / Nindividuals_S  #计算 Specificity
  Nsites_H <- ncol(otu_group_i)  #计算 Nsites H
  Nsites_SH <- apply(otu_group_i, 1, function(x) sum(x>0))  #计算 Nsites SH
  Occupancy <- Nsites_SH / Nsites_H  #计算 Occupancy
  spec_occu_group_i <- data.frame(layer = i, OTU = rownames(otu_group_i), 
                                  Specificity = Specificity, Occupancy = Occupancy, 
                                  Abundance_mean = rowMeans(otu_group_i), Taxonomy = otu_specif$Taxonomy)
  spec_occu <- rbind(spec_occu, spec_occu_group_i)  #合并各组统计
}
spec_occu <- spec_occu %>% mutate(layer = factor(layer, levels = c('SUR', 'SUB', 'PL')))
# 输出统计表格
# write.table(spec_occu, 'spec_occu.txt', row.names = FALSE, sep = '\t', quote = FALSE)
head(spec_occu)  # 该数据框包含各组中各 OTU 的名称、特异性、占有率、平均丰度、类群信息等
```
## 识别各组样本中的特化种（specialist species)
```{r}
# 在上述统计表格中直接根据特异性和占有率 ≥0.7 做筛选，保留下的 OTU 即为特化种
spec_occu_specialist <- subset(spec_occu, Specificity >= 0.7 & Occupancy >= 0.7) %>% 
  mutate(layer = factor(layer, levels = c('SUR', 'SUB', 'PL')))
head(spec_occu_specialist)
# 输出统计表格
# write.table(spec_occu_specialist, 'spec_occu_specialist.txt', row.names = FALSE, sep = '\t', quote = FALSE)
```
```{r, fig.width=6.5, fig.height=4.5}
#绘制 SPEC-OCCU 图, 添加阈值线
library(ggplot2)
p <- spec_occu %>% mutate(layer = factor(layer, levels = c('SUR', 'SUB', 'PL'))) %>%
  ggplot(aes(Occupancy, Specificity)) +
  # geom_point(aes(size = log10(Abundance_mean), color = 'Taxonomy'), alpha = 0.5) +
  geom_jitter(aes(size = log10(Abundance_mean), colour = Specificity >= 0.7 & Occupancy >= 0.7), alpha = 0.8) +  #如果觉得普通散点图中的点重叠严重不好看，可以仿照 Gweon et al (2021) 使用抖动点图来展示
  scale_colour_manual(values = c("grey", "#5f9ea0")) +
  scale_size(breaks = c(-1, -2, -3, -4), labels = c(expression(10^{-1}), expression(10^{-2}), expression(10^{-3}), expression(10^{-4})), range = c(0, 1)) +
  #scale_color_manual(values = c('#E7272E', '#F59D1F', '#768CC5', '#9BC648', '#794779', '#A19E9D'), limits = c('Proteobacteria', 'Actinobacteria', 'Acidobacteria', 'Bacteroidetes', 'Firmicutes', 'Others')) +
  theme(axis.title = element_text(colour = 'black'),
        axis.text = element_text(colour = 'black'),
        panel.grid = element_blank(), 
        panel.background = element_rect(fill = 'white', color = 'gray30'), 
        legend.key = element_blank(), legend.position = "bottom") +
  facet_wrap(~layer, ncol = 3) +
  scale_x_continuous(breaks = c(0, 0.5, 1), expand = c(0, 0), limit = c(0, 1)) +
  scale_y_continuous(breaks = c(0, 0.5, 1), expand = c(0, 0), limit = c(0, 1)) +
  labs(x = 'Occupancy', y = 'Specificity', size = 'Average relative abundance') + # add color option, color = 'Taxonomy'
  coord_cartesian(clip = 'off') +
  geom_segment(aes(x = 0.7, xend = 1, y = 0.7, yend = 0.7), linetype = 2)+
  geom_segment(aes(x = 0.7, xend = 0.7, y = 0.7, yend = 1), linetype = 2)

# p <- ggplot(spec_occu_specialist, aes(x = Occupancy, y = Specificity, colour = layer, size = log10(Abundance_mean))) +
#   geom_point(data = spec_occu, colour = "#0087A8", alpha = .2) +
#   geom_point() + 
#   scale_size(breaks = c(-1, -2, -3, -4), labels = c(expression(10^{-1}), expression(10^{-2}), expression(10^{-3}), expression(10^{-4})), range = c(0, 1)) +
#   theme(axis.title = element_text(colour = 'black'),
#         axis.text = element_text(colour = 'black'),
#         panel.grid = element_blank(), 
#         panel.background = element_rect(fill = 'white', color = 'gray30'), 
#         legend.key = element_blank(), legend.position = "bottom") +
#   facet_wrap(~layer, ncol = 3) +
#   scale_x_continuous(breaks = c(0, 0.5, 1), expand = c(0, 0), limit = c(0, 1)) +
#   scale_y_continuous(breaks = c(0, 0.5, 1), expand = c(0, 0), limit = c(0, 1)) +
#   labs(x = 'Occupancy', y = 'Specificity', size = 'Average relative abundance') + # add color option, color = 'Taxonomy'
#   coord_cartesian(clip = 'off') +
#   geom_segment(aes(x = 0.7, xend = 1, y = 0.7, yend = 0.7), color = "black", linetype = 2)+
#   geom_segment(aes(x = 0.7, xend = 0.7, y = 0.7, yend = 1), color = "black", linetype = 2)
# # save plot
# # cairo_pdf(filename = file.path(save.dir, "./figs/diff_abun/spec_occu_plot.pdf"), width = 8.5, height = 4.1, onefile = TRUE)
# # p
# # dev.off()
# # ggsave(file.path(save.dir, "./figs/diff_abun/spec_occu_plot.pdf"), p, width = 180, height = 90, units = "mm")
p
```
## tree图显示特化种的具体类群
```{r}
# devtools::install_github("ropensci/taxa", force = TRUE)
library(metacoder)
metacode_tax <- read.table(file.path(wd_16s, 'metacode_tax.txt'), header=T, sep="\t", comment.char="")
obj <- parse_tax_data(metacode_tax,
                      class_cols = "lineage", # the column that contains taxonomic information
                      class_sep = ";", # The character used to separate taxa in the classification
                      class_regex = "^(.+)__(.+)$", # Regex identifying where the data for each taxon is
                      class_key = c(tax_rank = "info", # A key describing each regex capture group
                                    tax_name = "taxon_name"))
# comparison between SUR, SUB and PL
# transform to proportion data
obj$data$tax_data <- calc_obs_props(obj, "tax_data")
spec_sur <- rownames(otu) %in% subset(spec_occu_specialist, layer == 'SUR')$OTU
spec_sub <- rownames(otu) %in% subset(spec_occu_specialist, layer == 'SUB')$OTU
spec_pl <- rownames(otu) %in% subset(spec_occu_specialist, layer == 'PL')$OTU
obj_sur <- filter_obs(obj, "tax_data", spec_sur, drop_taxa = T)
obj_sub <- filter_obs(obj, "tax_data", spec_sub, drop_taxa = T)
obj_pl <- filter_obs(obj, "tax_data", spec_pl, drop_taxa = T)
```
Getting per-taxon information
```{r}
obj_sur$data$tax_abund <- calc_taxon_abund(obj_sur, "tax_data", cols = metadata$sample_id)
obj_sub$data$tax_abund <- calc_taxon_abund(obj_sub, "tax_data", cols = metadata$sample_id)
obj_pl$data$tax_abund <- calc_taxon_abund(obj_pl, "tax_data", cols = metadata$sample_id)
# calculate the number of samples that have reads for each taxon
obj_sur$data$tax_occ <- calc_n_samples(obj_sur, "tax_abund", groups = metadata$layer, cols = metadata$sample_id)
obj_sub$data$tax_occ <- calc_n_samples(obj_sub, "tax_abund", groups = metadata$layer, cols = metadata$sample_id)
obj_pl$data$tax_occ <- calc_n_samples(obj_pl, "tax_abund", groups = metadata$layer, cols = metadata$sample_id)
```
Plotting taxonomic data
```{r, fig.align='center', fig.width=12, fig.height=5}
set.seed(1) # This makes the plot appear the same each time it is run 
p_sur <- obj_sur %>% 
  metacoder::filter_taxa(obj_sur$data$class_data$taxon_id[obj_sur$data$class_data$tax_rank == 'o'], 
                         supertaxa = TRUE) %>%
  heat_tree(node_label = taxon_names,
            node_size = n_obs,
            node_color = n_obs) # node_color_range = c("grey", "#F8766D")
# ggsave(file.path(save.dir, "./figs/diff_abun/spec_sur_tree.pdf"), p_sur, width = 89, height = 59, units = "mm")
set.seed(1) 
p_sub = obj_sub %>% 
  metacoder::filter_taxa(obj_sub$data$class_data$taxon_id[obj_sub$data$class_data$tax_rank == 'o'], 
                         supertaxa = TRUE) %>%
  heat_tree(node_label = taxon_names,
            node_size = n_obs,
            node_color = n_obs)
# ggsave(file.path(save.dir, "./figs/diff_abun/spec_sub_tree.pdf"), p_sub, width = 89, height = 59, units = "mm")
set.seed(1) 
p_pl = obj_pl %>% 
  metacoder::filter_taxa(obj_pl$data$class_data$taxon_id[obj_pl$data$class_data$tax_rank == 'o'], 
                         supertaxa = TRUE) %>%
  heat_tree(node_label = taxon_names,
            node_size = n_obs,
            node_color = n_obs)
# ggsave(file.path(save.dir, "./figs/diff_abun/spec_pl_tree.pdf"), p_pl, width = 89, height = 59, units = "mm")
cowplot::plot_grid(p_sur, p_sub, p_pl, ncol = 3)
# ggsave(file.path(save.dir, "./figs/diff_abun/spec_occu_tree.pdf"), spec_occu_tree, width = 210, height = 120, units = "mm")
```


